src/stim/simulators/tableau_simulator.test.cc

```diff
@@ -457,55 +458,56 @@ TEST(TableauSimulator, to_state_vector_canonical) {
-    }
-    for (size_t k = 0; k < 10; k++) {
-        auto state = Tableau<MAX_BITWORD_WIDTH>::random(4, SHARED_TEST_RNG());
-        ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 1}));
-        ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {2, 1}));
-        ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 1, 2, 3}));
-    }
-    {
-        auto state = Tableau<MAX_BITWORD_WIDTH>::random(12, SHARED_TEST_RNG());
-        ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 1, 2, 3}));
-        ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 10, 11}));
-        ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {11, 5, 7}));
-        ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}));
-    }
-}
-
-TEST(TableauSimulator, correlated_error) {
-    simd_bits<MAX_BITWORD_WIDTH> expected(5);
+})
+
+// template <size_t W>
+// bool vec_sim_corroborates_measurement_process(const Tableau<W> &state, const std::vector<uint32_t> &measurement_targets) {
+//     TableauSimulator<W> sim_tab(SHARED_TEST_RNG(), 2);
+//     sim_tab.inv_state = state;
+//     auto vec_sim = sim_tab.to_vector_sim();
+//     sim_tab.do_MZ(OpDat(measurement_targets));
+//     PauliString<W> buf(sim_tab.inv_state.num_qubits);
+//     size_t k = 0;
+//     for (auto t : measurement_targets) {
+//         buf.zs[t] = true;
+//         buf.sign = sim_tab.measurement_record.storage[k++];
+//         float f = vec_sim.project<W>(buf);
+//         if (fabs(f - 0.5) > 1e-4 && fabsf(f - 1) > 1e-4) {
+//             return false;
+//         }
+//         buf.zs[t] = false;
+//     }
+//     return true;
+// }
+//
+// TEST_EACH_WORD_SIZE_W(TableauSimulator, measurement_vs_vector_sim, {
+//     for (size_t k = 0; k < 10; k++) {
+//         auto state = Tableau<W>::random(2, SHARED_TEST_RNG());
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0}));
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {1}));
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 1}));
+//     }
+//     for (size_t k = 0; k < 10; k++) {
+//         auto state = Tableau<W>::random(4, SHARED_TEST_RNG());
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 1}));
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {2, 1}));
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 1, 2, 3}));
+//     }
+//     {
+//         auto state = Tableau<W>::random(12, SHARED_TEST_RNG());
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 1, 2, 3}));
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 10, 11}));
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {11, 5, 7}));
+//         ASSERT_TRUE(vec_sim_corroborates_measurement_process(state, {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}));
+//     }
+// })
```

Comment:
I intend to give this code further thought.

The issue here is `tableau_simulator.h` includes vector simulator. This is problematic for us here because `tableau_simulator.h` doesn't instantiate template member functions that `tableau_simulator.test` subsequently needs from vector_simulator.h. Once the test file gets round to including the vector simulator, all its template member function definitions are ifdef'd out.

I need to find a way to allow the `TableauSimulator` to coexist with the `VectorSimulator` in single translation unit.

Reply:
I don't think I understand what the issue is here. What's the actual compile error that occurs?

Reply:
```
/home/markturner289/code/Stim-fork/src/stim/simulators/tableau_simulator.test.cc: In instantiation of ‘bool vec_sim_corroborates_measurement_process(const stim::Tableau<W>&, const std::vector<unsigned int>&) [with long unsigned int W = 256]’:
/home/markturner289/code/Stim-fork/src/stim/simulators/tableau_simulator.test.cc:483:1:   required from here
/home/markturner289/code/Stim-fork/src/stim/simulators/tableau_simulator.test.cc:474:34: error: invalid operands of types ‘<unresolved overloaded function type>’ and ‘long unsigned int’ to binary ‘operator<’
  474 |         float f = vec_sim.project<W>(buf);
      |                   ~~~~~~~~~~~~~~~^~
/home/markturner289/code/Stim-fork/src/stim/simulators/tableau_simulator.test.cc: In instantiation of ‘bool vec_sim_corroborates_measurement_process(const stim::Tableau<W>&, const std::vector<unsigned int>&) [with long unsigned int W = 128]’:
/home/markturner289/code/Stim-fork/src/stim/simulators/tableau_simulator.test.cc:483:1:   required from here
/home/markturner289/code/Stim-fork/src/stim/simulators/tableau_simulator.test.cc:474:34: error: invalid operands of types ‘<unresolved overloaded function type>’ and ‘long unsigned int’ to binary ‘operator<’
/home/markturner289/code/Stim-fork/src/stim/simulators/tableau_simulator.test.cc: In instantiation of ‘bool vec_sim_corroborates_measurement_process(const stim::Tableau<W>&, const std::vector<unsigned int>&) [with long unsigned int W = 64]’:
/home/markturner289/code/Stim-fork/src/stim/simulators/tableau_simulator.test.cc:483:1:   required from here
/home/markturner289/code/Stim-fork/src/stim/simulators/tableau_simulator.test.cc:474:34: error: invalid operands of types ‘<unresolved overloaded function type>’ and ‘long unsigned int’ to binary ‘operator<’
make[2]: *** [CMakeFiles/stim_test_o3.dir/build.make:1987: CMakeFiles/stim_test_o3.dir/src/stim/simulators/tableau_simulator.test.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:86: CMakeFiles/stim_test_o3.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

Reply:
Finally worked it out.

Where previously, in `vec_sim_corroborates_measurement_process`, I had 

```C++
        float f = vec_sim.project<W>(buf);
```

I instead needed

```C++
        float f = vec_sim.template project<W>(buf);
```

Without `.template` the type of `vec_sim` was assumed to be a non-template.

I wasn't familiar with this syntax / theory, so found this blog post on dependant types fascinating: http://www.aerialmantis.co.uk/blog/2017/03/17/template-keywords/.